cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metadata;
+
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Abstract implementation for {@link ApplicationDetailFetcher} for common functionalities.
+ */
+public abstract class AbstractApplicationDetailFetcher implements ApplicationDetailFetcher {
+
+  /**
+   * Current implementation is suboptimal as it fetches each application individually either in
+   * separate db transaction or separate api call.
+   * TODO(samik) Add batch fetch by id in DefaultStore and expose via API
+   */
+  @Override
+  public Iterator<ApplicationDetail> getMulti(String namespace, Collection<String> appNames) {
+    Iterator<ApplicationReference> appIdIterator = appNames.stream()
+        .map(s -> new ApplicationReference(namespace, s)).iterator();
+    return new Iterator<ApplicationDetail>() {
+      @Override
+      public boolean hasNext() {
+        return appIdIterator.hasNext();
+      }
+
+      @Override
+      public ApplicationDetail next() {
+        try {
+          return get(appIdIterator.next());
+        } catch (NotFoundException | IOException e) {
+          throw new RuntimeException(e);
```

Comment:
@samdgupi to extend RuntimeException to add specific cases like NotFound

Reply:
Unfortunately the current exception class `ApplicationNotFoundException` is a checked Exception hence cannot be used here. Created a seperate exception `ApplictaionReferenceNotFoundException` for this usecase. Not a ideal scenario but this will help to identify the special case of notfound error

Reply:
this indicates an Iterator is not the best way to model this. Somebody using an iterator will not expect that they need to handle exceptions in the next() method.